snuba/web/rpc/v1/resolvers/R_eap_items/heatmap_builder.py

```diff
@@ -0,0 +1,291 @@
+        heatmap: HeatmapRequest,
+        in_msg: TraceItemStatsRequest,
+        routing_decision: RoutingDecision,
+        timer: Timer,
+    ) -> tuple[float | None, float | None, float | None]:
+        """
+        Returns the min value, max value, and bucket size for the y attribute.
+
+        example:
+        say we are building a heatmap for x attribute span.op and y attribute span.duration
+        we will get the min span.duration across all spans (items) and the max, then we will use
+        number_y_buckets to calculate the bucket size.
+        say min is 25, max is 350 and number_y_buckets is 4
+        then we will return (25, 350, 81.25)
+        """
+        x_attribute = heatmap.x_attribute
+        y_attribute = heatmap.y_attribute
+        y_attribute_val = attribute_key_to_expression(y_attribute)
+        filter = TraceItemFilter(
+            and_filter=AndFilter(
+                filters=[
+                    in_msg.filter,
+                    TraceItemFilter(exists_filter=ExistsFilter(key=x_attribute)),
+                    TraceItemFilter(exists_filter=ExistsFilter(key=y_attribute)),
+                ]
+            )
+        )
+        filter_expression = trace_item_filters_to_expression(
+            filter,
+            (attribute_key_to_expression),
+        )
+        condition = base_conditions_and(in_msg.meta, filter_expression)
+        min_max_query = Query(
+            from_clause=EAP_ITEMS_ENTITY,
+            selected_columns=[
+                SelectedExpression(name="max_val_y", expression=f.max(y_attribute_val)),
+                SelectedExpression(name="min_val_y", expression=f.min(y_attribute_val)),
+            ],
+            condition=condition,
+        )
+        treeify_or_and_conditions(min_max_query)
+        snuba_request = self._build_snuba_request(min_max_query)
+        min_max_res = (
+            run_query(
+                dataset=PluggableDataset(name="eap", all_entities=[]),
+                request=snuba_request,
+                timer=timer,
+            )
+            .result["data"][0]
+            .values()
+        )
+        if list(min_max_res) == [None, None]:
+            return (None, None, None)
+        min_val_y = min(min_max_res)
+        max_val_y = max(min_max_res)
+        group_size_y = (max_val_y - min_val_y) / heatmap.num_y_buckets
+        return (min_val_y, max_val_y, group_size_y)
+
+    def _get_y_buckets(
+        self, min_y: float, max_y: float, group_size_y: float
+    ) -> list[AttributeValue]:
+        buckets = [
+            AttributeValue(
+                val_str=f"[{min_y + i * group_size_y}, {min_y + (i + 1) * group_size_y})",
+            )
+            for i in range(self.heatmap.num_y_buckets)
+        ]
+        buckets[-1].val_str = buckets[-1].val_str[:-1] + "]"
+        return buckets
+
+    def _build_heatmap_query(
+        self,
+        min_val_y: float,
+        max_val_y: float,
+        group_size_y: float,
+    ) -> Query:
```

Comment:
This query is a little hard to follow, can you explain inline how what the sql you're generating is? If I were to try and fix a bug in this it would take me a fair bit of time to decipher it